### PR TITLE
Adds Config.md

### PR DESCRIPTION
**Pull request by bluemonster122**
Similar to #130 solves #119

### CODE REVIEW

File: docs/config/@config.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+| Property|Type       | Default Value |Comment 
+|       -:|:-         |       :-:     |:-
+|    value|`String[]` |       N/A     |This can be passed as a `String` or `String[]`, a `String[]` will form a multi-line comment where 1 element = 1 line.
+
+### Example
+```java
+@Comment({
+  "You can add comments using this",
+  "and if you supply an array it will me multiline"
```
Comment (TehNut) [inline docs/config/@config.md]: me -> be

File: docs/config/@config.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+})
+public static boolean doTheThing = true;
+```
+This will produce a `boolean` config value with name `doTheThing`, the defualt value `true` and the comment provided in the config file.
+
+@Name Use
+---------
+
+You should use the `@Name` annotation to give a config a user friendly name in the config file. This has 1 property:
+| Property | Type    | Default Value |
```
Comment (TehNut) [inline docs/config/@config.md]: Need a newline between 68 and 69. The formatting is currently broken.

File: docs/config/@config.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+@RangeDouble(min = 0, max = Math.PI) 
+public static double chanceToDrop = 2;
+```
+This will produce a `double` config value with the name `chanceToDrop`, and the default value `Math.PI` in the config file; it will ensure the value stays between the bounds of `0` and `Math.PI`.
+
+@LangKey Use
+------------
+
+You should use the `@LangKey` annotation to give a config a user friendly via .lang files in the mod options menu. NOTE: This does not affect anything in the config file itself. This has 1 property:
+| Property | Type    | Default Value |
```
Comment (TehNut) [inline docs/config/@config.md]: Need a newline between 118 and 119. The formatting is currently broken.

File: docs/config/@config.md [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+* [@RangeInt Use][rangeintuse]
+* [@RangeDouble Use][rangedoubleuse]
+* [@LangKey Use][langkeyuse]
+* [@RequiresMcRestart Use][requiresmcrestartuse]
+* [@RequiresWorldRestart Use][requiresworldrestartuse]
+
+Basics
+------
+
+A Config should contain static fields, one per config option. There are a few exceptions, such as [Nested Configs][nestedconfigs] and [Sub Categories][subcategories]. Note that you can use an enum as a config field. To add information to and control config values, you can use the plethora of annotations provided in the `@Config` class.
```
Comment (TehNut) [inline docs/config/@config.md]: What are you defining as the difference between a "Nested Config" and a "Sub Category". You appear to be doing the same thing in both.

File: docs/config/@config.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+```
+
+@Comment Use
+------------
+
+Adding a `@Comment` annotation to a field will add a comment to the config in it's file. It has 1 property:
+
+| Property|Type       | Default Value |Comment 
+|       -:|:-         |       :-:     |:-
+|    value|`String[]` |       N/A     |This can be passed as a `String` or `String[]`, a `String[]` will form a multi-line comment where 1 element = 1 line.
```
Comment (elifoster) [inline docs/config/@config.md]: This is a comma splice

Comment (elifoster) [inline docs/config/@config.md]: The Type column and the Comment are contradictory. Do you have to pass a `String[]` even if it's one line (`@Comment({"Comment"})`), or can you pass a `String` for a single line comment (`@Comment("Comment")`)?

File: docs/config/@config.md (hunk 1 above)
Comment (elifoster) [inline docs/config/@config.md]: In the Comment column above, you use `multi-line`, but here you use `multiline`.

File: docs/config/@config.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+### Example
+```java
+@Comment({
+  "You can add comments using this",
+  "and if you supply an array it will me multiline"
+})
+public static boolean doTheThing = true;
+```
+This will produce a `boolean` config value with name `doTheThing`, the defualt value `true` and the comment provided in the config file.
```
Comment (elifoster) [inline docs/config/@config.md]: This is a really wordy sentence. I'm not sure how I'd reword it, but I think it could use some work.

File: docs/config/@config.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+@RangeDouble Use
+----------------
+
+You should use `@RangeDouble` to limit a `double` config value. It has a couple of properties:
+
+| Property|Type   |    Default Value   |
+|       -:|:-     |        :-:         |
+|      min|double | `Double.MIN_VALUE` |
+|      max|double | `Double.MAX_VALUE` |
+
```
Comment (elifoster) [inline docs/config/@config.md]: I see we have `@RangeInt` and `@RangeDouble`, do we have `@RangeFloat`, `@RangeLong`, etc.? If not that should be explicitly stated somewhere.

File: docs/config/@config.md [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+### Example
+```java
+@RangeDouble(min = 0, max = Math.PI) 
+public static double chanceToDrop = 2;
```
Comment (elifoster) [inline docs/config/@config.md]: Can we use `Double` or do we have to use the primitive type?

File: docs/config/@config.md [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+```java
+@Name("FE/T for the thing")
+public static int thingFE = 50; 
+```
+This will produce an `int` config value with the name `FE/T for the thing`, and the default value `50` in the config file.
+
+@RangeInt Use
+-------------
+
+You should use `@RangeInt` to limit an `int` config value. It has a couple of properties:
```
Comment (elifoster) [inline docs/config/@config.md]: Can we use `Integer` or do we have to use the primitive type?

File: docs/config/@config.md [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+```java
+@RangeDouble(min = 0, max = Math.PI) 
+public static double chanceToDrop = 2;
+```
+This will produce a `double` config value with the name `chanceToDrop`, and the default value `Math.PI` in the config file; it will ensure the value stays between the bounds of `0` and `Math.PI`.
+
+@LangKey Use
+------------
+
+You should use the `@LangKey` annotation to give a config a user friendly via .lang files in the mod options menu. NOTE: This does not affect anything in the config file itself. This has 1 property:
```
Comment (elifoster) [inline docs/config/@config.md]: `You should use the `@LangKey` annotation to give a config a user friendly via .lang files in the mod options menu.` does not make sense

Comment (elifoster) [inline docs/config/@config.md]: `NOTE: ` is unnecessary imo

File: docs/config/@config.md [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+  public static NestedConfig nestedConfig = new NestedConfig();
+  
+  public static class NestedConfig {
+    public String message = "Don't go that way";
+  }
+}
+```
+
+Sub Categories
+--------------
```
Comment (elifoster) [inline docs/config/@config.md]: What exactly is the difference between a Sub Category and a Nested config? They look pretty much the same to me

File: docs/config/@config.md [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+### Example
+```java
+@RequiresWorldRestart
+public static boolean someOtherworldlyThing = false;
+```
+This will force the world to be restarted if the config is changed in the mod options menu.
+
+Nested Configs
+--------------
+A Nested config is possible like this:
```
Comment (elifoster) [inline docs/config/@config.md]: Please explain _what_ a Nested Config is not just how to make one

File: mkdocs.yml [hunk 13]
```diff
@@ -33,6 +33,7 @@ pages:
     - Capabilities: 'datastorage/capabilities.md'
     - World Saved Data: 'datastorage/worldsaveddata.md'
     - Extended Entity Properties: 'datastorage/extendedentityproperties.md'
+    - Config: 'config/@config.md'
```
Comment (elifoster) [inline mkdocs.yml]: If possible the name should be changed to `@Config` because ideally `Configuration` would also be documented somewhere and it could be confusing.

File: docs/config/@config.md [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+-----------
+
+This annotation is used to denote a class is the container for some configuration options. There are 4 properties:
+
+|     Property|Type     | Default Value | Comment 
+|           -:|:-       |     :-:       |:-
+|        modid|`String` |     N/A       | The mod id that this configuration is associated with. 
+|         name|`String` |     `""`      | A user friendly name for the config file, the default will be modid. 
+|         type|`Type`   |`Type.INSTANCE`| The type this is, right now the only value is `Type.INSTANCE`. This is intended to be expanded upon later for more Forge controlled configs. 
+|     category|`String` |  `"general"`  | Root element category, defaults to "general", if this is an empty string then the root category is disabled. Any primitive fields will cause an error, and you must specify sub-category objects.
```
[the remark this answers is not in the document]
Comment (TehNut) [reply docs/config/@config.md]: Is it even worth mentioning that there? It's already in the column to the left.

[the remark this answers is not in the document]
Comment (elifoster) [reply docs/config/@config.md]: Oh yeah, you're right.

File: docs/config/@config.md (hunk 7 above)
Comment (TehNut) [reply docs/config/@config.md]: We do not.

**Comment by bluemonster122** [comment]
This will give me something to do on the way to uni tomorrow :D

File: docs/config/@config.md (hunk 11 above)
Comment (bluemonster122) [reply docs/config/@config.md]: I assumed a difference, which didn't exist...  What an ass that made me look >.>

File: mkdocs.yml (hunk 13 above)
Comment (bluemonster122) [reply mkdocs.yml]: If it breaks it's not my fault 😛

File: docs/config/@config.md [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+* [@RequiresMcRestart Use][requiresmcrestartuse]
+* [@RequiresWorldRestart Use][requiresworldrestartuse]
+
+Basics
+------
+
+A class annotated with `@Config` will have any fields turned into config options. Said fields can be annotated to add information, using the plethora of annotations provided in the `@Config` class.
+
+!!! note
+  Enum config values generate a comment that is as many lines long as the enum has values.
```
Comment (TehNut) [inline docs/config/@config.md]: newline between 24 and 25. Also, it's 4 spaces, not 2.

File: docs/config/@config.md [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+------------
+
+Adding a `@Comment` annotation to a field will add a comment to the config in its file. 
+
+It has 1 property:
+
+| Property|Type       | Default Value |Comment 
+|       -:|:-         |       :-:     |:-
+|    value|`String[]` |       N/A     |A `String[]` will form a multi-line comment where 1 element = 1 line.
+!!! note
```
Comment (TehNut) [inline docs/config/@config.md]: Newline between 55 and 56. Formatting is very broken.

image

File: docs/config/@config.md [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+})
+public static boolean doTheThing = true;
+```
+This would produce the following config:
+```
+# You can add comments using this
+# and if you supply an array it will me multiline
+B:doTheThing=true   
+``` 
+This will produce a `boolean` config value with name `doTheThing`, the defualt value `true` and the comment provided in the config file.
```
Comment (TehNut) [inline docs/config/@config.md]: This line is unnecessary with the example provided above. Also, typo: `defualt`

File: docs/config/@config.md [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+
+@Name Use
+---------
+
+You should use the `@Name` annotation to give a config a user friendly name in the config file.
+
+This has 1 property:
+| Property | Type    | Default Value |
```
Comment (TehNut) [inline docs/config/@config.md]: Newline between 80 and 81. Formatting is broken.

File: docs/config/@config.md [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+```
+I:"FE/T for the thing"=50
+```
+
+@RangeInt Use
+-------------
+
+You should use `@RangeInt` to limit an `int` or `Integer` config value. 
+
+It has a couple of properties:
```
Comment (TehNut) [inline docs/config/@config.md]: Explicitly state 2 instead of "a couple". You do this throughout the above types.

File: docs/config/@config.md [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+# Max: 2147483647
+I:thingFECapped=50
+```
+
+@RangeDouble Use
+----------------
+
+You should use `@RangeDouble` to limit a `double` or `Double` config value. 
+
+It has a couple of properties:
```
Comment (TehNut) [inline docs/config/@config.md]: Explicitly state 2 instead of "a couple". You do this throughout the above types.

File: docs/config/@config.md [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+```
+This will produce the following config:
+```
+# Min: 0.0
+# Max: 3.141592653589793
+D:chanceToDrop=2.0
+```
+
+!!! note
+  There is not currenty (in 1.12.2) a `@RangedFloat` or `@RangedLong` or any other variant of the Ranged values.
```
Comment (TehNut) [inline docs/config/@config.md]: newline between 24 and 25. Also, it's 4 spaces, not 2. 

Misspelled "currently".

File: docs/config/@config.md [hunk 22]
```diff
@@ -0,0 +1,222 @@
...
+!!! note
+  There is not currenty (in 1.12.2) a `@RangedFloat` or `@RangedLong` or any other variant of the Ranged values. 
+
+@LangKey Use
+------------
+
+If you want to add translations for your configs in the mod options menu, add to the config's field `@LangKey`.
+
+This has 1 property:
+| Property | Type    | Default Value |
```
Comment (TehNut) [inline docs/config/@config.md]: Newline between 152 and 153. Formatting is broken.

File: docs/config/@config.md [hunk 23]
```diff
@@ -0,0 +1,222 @@
...
+### Example
+```java
+@RequiresWorldRestart
+public static boolean someOtherworldlyThing = false;
+```
+This will force the world to be restarted if the config is changed in the mod options menu.
+
+Sub Categories
+--------------
+If you want Sub Categories, you must specify them. A good way to do this is to make a private inner class, and create a singleton instance in it's parent.
```
Comment (TehNut) [inline docs/config/@config.md]: >If you want Sub Categories, you must specify them.

This could probably be removed. If you want *anything* you must specify it. No need to state that here.

File: docs/config/@config.md [hunk 24]
```diff
@@ -180,7 +186,7 @@ This will force the world to be restarted if the config is changed in the mod op
 
 Sub Categories
 --------------
-If you want Sub Categories, you must specify them. A good way to do this is to make a private inner class, and create a singleton instance in it's parent. 
+A good way to male a `Sub Category` is to make a private inner class, and create a singleton instance in it's parent.
```
Comment (TehNut) [inline docs/config/@config.md]: male -> make

Why is "Sub Category" in a code block?

Honestly I'm not quite sure how I feel about this sentence as a whole. It seems more like a personal preference, since it can just be any object type.

Comment (bluemonster122) [reply docs/config/@config.md]: I see your point, i'll change it real quick.... Thank god you guys are so good at spotting mistakes... i cant spot my mistakes ever.

File: docs/config/@config.md [hunk 25]
```diff
@@ -0,0 +1,228 @@
...
+### Example
+```java
+@RequiresWorldRestart
+public static boolean someOtherworldlyThing = false;
+```
+This will force the world to be restarted if the config is changed in the mod options menu.
+
+Sub Categories
+--------------
+To make a Sub Category make an object, then add it as a static field in the parent categories class. The object's non-static fields will become configs in that Sub Category.
```
Comment (TehNut) [inline docs/config/@config.md]: non-static -> member

File: docs/config/@config.md [hunk 26]
```diff
@@ -0,0 +1,228 @@
...
+
+An example of how to setup a Sub Category:
```
Comment (TehNut) [inline docs/config/@config.md]: This example is a bit large for just covering how sub categories work. We've already gone over `@RangeInt` and `@Comment`. Maybe just something basic like this?

```java
@Config(modid = "modid")
public class Configs {
  public static SubCategory subcat = new SubCategory();

  private static class SubCategory {
    public boolean someBool; 
  }
}
```

File: docs/config/@config.md [hunk 27]
```diff
@@ -186,32 +186,27 @@ This will force the world to be restarted if the config is changed in the mod op
...
 
   private static class SubCategory {
-    @Config.RangeInt(min = 0, max = 1000)
-    @Config.Comment("This is the FE/t used by the thing")
-    public int feForTheThing; 
+    public boolean someBool; 
 
-    public SubCategory(int defaultFE) {
-      feForTheThing = defaultFE;
+    public SubCategory(int someBool) {
```
Comment (TehNut) [inline docs/config/@config.md]: This constructor is still unnecessary fluff. You can define a default value by setting the field's value. bools also default to false, so it isn't even required here.

Comment (bluemonster122) [reply docs/config/@config.md]: ok :)

Comment (TehNut) [reply docs/config/@config.md]: Also I just realized you left `int` in that example constructor lol

**Comment by bluemonster122** [comment]
You are doing good things... I'm bad at writing

File: docs/config/@config.md (hunk 27 above)
Comment (bluemonster122) [reply docs/config/@config.md]: yup, i just woke up so i'm not completely with it... but idk if i ever am....

File: docs/config/@config.md [hunk 28]
```diff
@@ -0,0 +1,219 @@
...
+
+It has 1 property:
+
+| Property|Type       | Default Value |Comment 
+|       -:|:-         |       :-:     |:-
+|    value|`String[]` |       N/A     |A `String[]` will form a multi-line comment where 1 element = 1 line.
+
+!!! note
+
+    The value in this annotation can be passed as a `String` because java is cool and turns it into a `String[]`.
```
Comment (elifoster) [inline docs/config/@config.md]: This is weird. Instead, change the Type in the table above to `String[]`/`String` and clarify this in in the Comment section by saying "A `String[]` will form a multi-line comment where 1 element is 1 line. A single `String` can be passed to form a single-line comment."

File: docs/config/@config.md [hunk 29]
```diff
@@ -0,0 +1,219 @@
...
+@Comment({
+  "You can add comments using this",
+  "and if you supply an array it will be multi-line"
+})
+public static boolean doTheThing = true;
+```
+This would produce the following config:
+```
+# You can add comments using this
+# and if you supply an array it will me multiline
```
Comment (elifoster) [inline docs/config/@config.md]: This comment does not match the value provided in the code example. Specifically, `multiline` should be `multi-line`, but I would just copy them directly to make sure there are no other typos.

File: docs/config/@config.md [hunk 30]
```diff
@@ -0,0 +1,219 @@
...
+### Example
+```java
+@RequiresWorldRestart
+public static boolean someOtherworldlyThing = false;
+```
+This will force the world to be restarted if the config is changed in the mod options menu.
+
+Sub Categories
+--------------
+To create a Sub Category make an object, then add it as a static field in the parent category's class. The object's member fields will become configs in that Sub Category.
```
Comment (elifoster) [inline docs/config/@config.md]: What is a Sub Category and why/when should you use them?

**Comment by TehNut** [comment]
@elifoster line 150 includes a note about other range types.

**Comment by bluemonster122** [comment]
I just saw that the docs are not supposed to have functional copy/paste-able code... should I remove the examples?

**Comment by tterrag1098** [comment]
I don't mind the examples here. This seems pretty good to me.
